### PR TITLE
Draft: nvidia-enabled Docker container and scripting

### PR DESCRIPTION
**Pull request by mmguero**
related to s0md3v/roop#171

Adds `docker` directory to source tree containing:

* `Dockerfile` - defining the Docker container based on `nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04`
* `docker-compose.yml` - mainly just for testing as it prints out the output of `nvidia-smi` and that's it
* `build_roop_docker.sh` - convenience script for building the image
* `docker-uid-gid-setup.sh` - entrypoint for dropping privs from root to non-root UID
* `roop-docker.sh` - wrapper script for `docker run` to massage and pass some of the commands through

My Linux host environment has the NVIDIA driver and NVIDIA [container runtime](https://developer.nvidia.com/nvidia-container-runtime). I build the image and run my wrapper script like this:

```bash
$ ./docker/build_roop_docker.sh 
[+] Building 2.8s (13/13) FINISHED                                                                                                                                                             
 => [internal] load .dockerignore                                                                                                                                                         0.0s
 => => transferring context: 85B                                                                                                                                                          0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                      0.0s
 => => transferring dockerfile: 3.18kB                                                                                                                                                    0.0s
 => [internal] load metadata for docker.io/nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04                                                                                                  1.3s
 => [internal] load build context                                                                                                                                                         0.0s
 => => transferring context: 13.31kB                                                                                                                                                      0.0s
 => [1/8] FROM docker.io/nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04@sha256:334145928724f859d419359ce35dc77aac99c6ea8590ddcb5e52eb6a532ac963                                            0.0s
 => CACHED [2/8] COPY --chmod=644 ./inswapper_128.onnx /roop/inswapper_128.onnx                                                                                                           0.0s
 => CACHED [3/8] COPY ./requirements.txt /tmp/requirements.txt                                                                                                                            0.0s
 => CACHED [4/8] RUN apt-get -q update &&     apt-get install -q -y --no-install-recommends         build-essential         curl         ffmpeg         procps         psmisc         py  0.0s
 => CACHED [5/8] COPY --chmod=755 ./*.py /roop/                                                                                                                                           0.0s
 => [6/8] COPY --chmod=755 ./roop/*.py /roop/roop                                                                                                                                         0.0s
 => [7/8] COPY --chmod=755 ./docker/docker-uid-gid-setup.sh /usr/local/bin/docker-uid-gid-setup.sh                                                                                        0.0s
 => [8/8] WORKDIR /roop                                                                                                                                                                   0.0s
 => exporting to image                                                                                                                                                                    1.4s
 => => exporting layers                                                                                                                                                                   1.4s
 => => writing image sha256:c374d267a287da96c78a8984046390b655b8d3e5f8de8e1445ec7138516201ad                                                                                              0.0s
 => => naming to docker.io/library/roop:latest         

$ ./docker/roop-docker.sh -r nvidia -f ~/tmp/face.jpg -t ~/tmp/target.mp4 -o ~/tmp/outdocker.mp4 -- --gpu-vendor=nvidia --gpu-threads=2
rooper
uid=1001(rooper) gid=1001(rooper) groups=1001(rooper),5(tty),27(sudo)
...
Applied providers: ['CUDAExecutionProvider', 'CPUExecutionProvider'], ...
...
Video saved as: /data/outdocker.mp4 

Status: swap successful!
/home/user/tmp/outdocker.mp4
```

If this is useful to you, even as starting point, feel free to pull and adjust as you see fit. Or if you don't want to accept it feel free to close, no problem either way.

### CODE REVIEW

**Comment by mmguero** [comment]
Note that my this docker image is huge, like 16GB because of all the dependencies it installs.

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+COPY ./requirements.txt /tmp/requirements.txt
+
+RUN apt-get -q update && \
+    apt-get install -q -y --no-install-recommends \
+        build-essential \
+        curl \
+        ffmpeg \
+        procps \
+        psmisc \
+        python3 \
```
Comment (henryruhs) [inline docker/Dockerfile:29]: Why not another `FROM python-3.9-slim` ? I do not plan to maintain all this

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+ENV DEBIAN_FRONTEND noninteractive
+ENV PYTHONUNBUFFERED 1
+ENV PYTHONDONTWRITEBYTECODE 1
+
+ENV BUFFALO_L_URL "https://github.com/deepinsight/insightface/releases/download/v0.7/buffalo_l.zip"
+# ENV INSWAPPER_128_URL "https://1drv.ms/u/s!AsHA3Xbnj6uAgxhb_tmQ7egHACOR?e=CPoThO"
+
+COPY --chmod=644 ./inswapper_128.onnx /roop/inswapper_128.onnx
+COPY ./requirements.txt /tmp/requirements.txt
```
Comment (henryruhs) [inline docker/Dockerfile:20]: Why not checking out the repo here instead? It is not a good idea to create a container that only works within a working copy of the project

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      usermod -a -G sudo ${PUSER} && \
+      echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers && \
+    python3 -m pip install --index-url https://download.pytorch.org/whl/cu118 torch torchvision torchaudio && \
+        python3 -m pip install onnxruntime-gpu && \
+        python3 -m pip install -r /tmp/requirements.txt && \
+    ln -s -r /usr/local/cuda-11.8/targets/x86_64-linux/lib/libnvrtc.so.11.2 /usr/local/cuda-11.8/targets/x86_64-linux/lib/libnvrtc.so.11 && \
+        ln -s -r /usr/local/cuda-11.8/targets/x86_64-linux/lib/libnvrtc.so.11 /usr/local/cuda-11.8/targets/x86_64-linux/lib/libnvrtc.so && \
+    mkdir -p /roop/roop /home/${PUSER}/.insightface/models/buffalo_l && \
+        # curl -fsSL -o /roop/inswapper_128.onnx "${INSWAPPER_128_URL}" && \
+        curl -fsSL -o /tmp/buffalo_l.zip "${BUFFALO_L_URL}" && \
```
Comment (henryruhs) [inline docker/Dockerfile:54]: there is no need to download buffalo in your own

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        unzip && \
+    apt-get -y -q --allow-downgrades --allow-remove-essential --allow-change-held-packages autoremove && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+COPY --chmod=755 ./*.py /roop/
+COPY --chmod=755 ./roop/*.py /roop/roop
+COPY --chmod=755 ./docker/docker-uid-gid-setup.sh /usr/local/bin/docker-uid-gid-setup.sh
+
+ENV NVIDIA_VISIBLE_DEVICES all
```
Comment (henryruhs) [inline docker/Dockerfile:72]: Basic container should be nvidia independend

File: docker/build_roop_docker.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
```
Comment (henryruhs) [inline docker/build_roop_docker.sh:1]: I don't understand why we need this

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+version: '3.9'
```
Comment (henryruhs) [inline docker/docker-compose.yml:1]: a base docker-compose.yml should be nvidia-free ... another like docker-compose.cuda.yml can be nvidia based ... you can combine them running docker-compose

File: docker/docker-uid-gid-setup.sh [hunk 7]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
```
Comment (henryruhs) [inline docker/docker-uid-gid-setup.sh:1]: this will not land in our repo - too much scripting going on here

File: docker/roop-docker.sh [hunk 8]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
```
Comment (henryruhs) [inline docker/roop-docker.sh:1]: will also not land in our repo... I don't understand most of it and whatever it does can be done via a shared volume to the host I guess

**Comment by henryruhs** [comment]
Thanks for the contribution. For me this goes a bit over the top, I like to have simple solutions that I can actual maintain.

**Pull request closed without merging**